apps/server/src/modules/tool/service/external-tool.service.ts

```diff
@@ -1,16 +1,94 @@
+import { ExternalToolServiceMapper } from './mapper/external-tool-service.mapper';
 
 @Injectable()
 export class ExternalToolService {
-	constructor(private externalToolRepo: ExternalToolRepo) {}
+	constructor(
+		private readonly externalToolRepo: ExternalToolRepo,
+		private readonly oauthProviderService: OauthProviderService,
+		private readonly mapper: ExternalToolServiceMapper,
+		private readonly schoolExternalToolRepo: SchoolExternalToolRepo,
+		private readonly courseExternalToolRepo: CourseExternalToolRepo,
+		@Inject(DefaultEncryptionService) private readonly encryptionService: IEncryptionService
+	) {}
 
 	async createExternalTool(externalToolDO: ExternalToolDO): Promise<ExternalToolDO> {
-		const externalTool: ExternalToolDO = await this.externalToolRepo.save(externalToolDO);
-		return externalTool;
+		if (externalToolDO.config instanceof Lti11ToolConfigDO) {
+			externalToolDO.config.secret = this.encryptionService.encrypt(externalToolDO.config.secret);
+		} else if (externalToolDO.config instanceof Oauth2ToolConfigDO) {
+			const oauthClient: ProviderOauthClient = this.mapper.mapDoToProviderOauthClient(
+				externalToolDO.name,
+				externalToolDO.config
+			);
+
+			await this.oauthProviderService.createOAuth2Client(oauthClient);
+		}
+
+		const created: ExternalToolDO = await this.externalToolRepo.save(externalToolDO);
+		return created;
+	}
+
+	async findExternalTools(
+		query: Partial<ExternalToolDO>,
+		options: IFindOptions<ExternalToolDO>
+	): Promise<Page<ExternalToolDO>> {
+		const tools: Page<ExternalToolDO> = await this.externalToolRepo.find(query, options);
+		tools.data = await Promise.all(
+			tools.data.map(async (tool: ExternalToolDO): Promise<ExternalToolDO> => {
+				if (tool.config instanceof Oauth2ToolConfigDO) {
+					const oauthClient: ProviderOauthClient = await this.oauthProviderService.getOAuth2Client(
+						tool.config.clientId
+					);
+					this.applyProviderOauthClientToDO(tool.config, oauthClient);
+				}
+				return tool;
+			})
+		);
+
+		return tools;
+	}
+
+	async findExternalToolById(id: EntityId): Promise<ExternalToolDO> {
+		const tool: ExternalToolDO = await this.externalToolRepo.findById(id);
+
+		if (tool.config instanceof Oauth2ToolConfigDO) {
+			const oauthClient: ProviderOauthClient = await this.oauthProviderService.getOAuth2Client(tool.config.clientId);
+			this.applyProviderOauthClientToDO(tool.config, oauthClient);
+		}
+
+		return tool;
+	}
+
+	async deleteExternalTool(toolId: string): Promise<void> {
+		const schoolExternalTools: SchoolExternalToolDO[] = await this.schoolExternalToolRepo.findByToolId(toolId);
+		const schoolExternalToolIds: string[] = schoolExternalTools.map(
+			(schoolExternalTool: SchoolExternalToolDO): string => {
+				// We can be sure that the repo returns the id
+				return schoolExternalTool.id as string;
+			}
+		);
+
+		await Promise.all([
+			this.courseExternalToolRepo.deleteBySchoolExternalToolIds(schoolExternalToolIds),
```

Comment:
We should not delete domain objects by id. See comment in `BaseDORepo`.

Reply:
Reference.createFromPK does not work properly when an entity class that has private methods. Typescript is displaying errors that `getShortTitle` is missing when private is used.

https://github.com/hpi-schul-cloud/schulcloud-server/blob/2041e7e01c5d946683e7bc16dfce6119ac5bb83a/apps/server/src/shared/repo/courseexternaltool/course-external-tool.repo.ts#L54-L54